src/ansys/fluent/core/session.py

```diff
@@ -343,8 +306,93 @@ def download(self, file_name: str, local_file_path: Optional[str] = None):
         FileNotFoundError
             If the remote file does not exist.
         """
-        if self.file_service:
-            if self.file_service.file_exist(file_name):
-                self.file_service.download_file(file_name, local_file_path)
+        file_service = self._file_service()
+        if file_service:
+            if file_service.file_exist(file_name):
+                file_service.download_file(file_name, local_file_path)
             else:
                 raise FileNotFoundError("Remote file does not exist.")
+
+    def _wait_for_file(self, file_name: str):
+        """Wait for file to get ready for upload or download.
+
+        Parameters
+        ----------
+        file_name : str
+            Case file name
+        Raises
+        ------
+        FileNotFoundError
+            If a case file does not exist.
+        """
+        file_service = self._file_service()
+        start_time = time.time()
+        max_wait_time = 100
+        while (time.time() - start_time) < max_wait_time:
+            if file_service.file_exist(os.path.basename(file_name)):
+                break
+            max_wait_time -= 1
+            time.sleep(3)
+        else:
+            raise FileNotFoundError(f"{file_name} does not exist.")
+
+    def _pypim_upload_helper(self, file_name, is_meshing, api):
+        """Uploads a case file if not available on the server.
+
+        Parameters
+        ----------
+        file_name : str
+            Case file name
+        is_meshing: bool
+            True if mode is meshing, otherwise False
+        api: Session object property
+            either session.tui or session.file
+        Raises
+        ------
+        FileNotFoundError
+            If a case file does not exist.
+        """
+        file_service = self._file_service()
+        if os.path.isfile(file_name):
+            if not file_service.file_exist(os.path.basename(file_name)):
+                self.upload(file_name)
+                self._wait_for_file(file_name)
+        elif file_service.file_exist(os.path.basename(file_name)):
+            pass
+        else:
+            raise FileNotFoundError(f"{file_name} does not exist.")
+        if is_meshing:
+            api(os.path.basename(file_name))
+        else:
+            api(file_name=os.path.basename(file_name))
```

Comment:
@hpohekar 

This part of the design prevents this utility from being reusable. 

Any differences between clients of the utility need to be placed _behind the abstraction_.
So, it means that we need each client to pass in a simple callable, and if one client says _I need to use a named argument_ then he must achieve that inside his own implementation. You are almost there - you just need this one final adjustment!

Reply:
Ok. I will try to handle this. Thank you.

Reply:
ok.